server/controllers/finance/reports/incomeExpense/index.js

```diff
@@ -17,282 +17,141 @@ const _ = require('lodash');
-
-  return db.exec(query, [accountId, dateFrom, dateTo, accountId, dateFrom, dateTo]);
+function fetchIncomeExpense(options) {
+    let result = {};
+    return fetchIncome(options)
+        .then((incomes) => {
+            _.merge(result, incomes);
+            return fetchExpense(options);
+        })
+        .then((expenses) => {
+            _.merge(result, expenses);
+            return db.one(`SELECT IFNULL(${result.incomeAggregation.balance} - ${result.expenseAggregation.balance}, 0) AS finalBalance`);
+        })
+        .then((finalAggregate) =>{
+            result.isLost = finalAggregate.finalBalance <= 0;            
+            _.merge(result, {finalBalance : finalAggregate.finalBalance <= 0 ? finalAggregate.finalBalance * -1 : finalAggregate.finalBalance });
+            // it is an incomeExpense report            
+            result.isIncomeExpense = true;
+            return result;
+        });
 }
 
-
-/**
- * @function document
- * @description process and render the incomeExpense report document
- */
-function document(req, res, next) {
-  const session = {};
-  const params = req.query;
-
-  let documentReport;
-  let sumIncome = 0;
-  let sumExpense = 0;
-
-  session.dateFrom = params.dateFrom;
-  session.dateTo = params.dateTo;
-  session.reportType = params.reportType;
-
-  _.defaults(params, { orientation : 'landscape', user : req.session.user });
-
-  try {
-    documentReport = new ReportManager(TEMPLATE, req.session, params);
-  } catch (e) {
-    next(e);
-    return;
-  }
-
-  processingIncomeExpenseReport(params)
-    .then(incomeExpense => {
-      incomeExpense.reportIncome = false;
-      incomeExpense.reportExpense = false;
-      incomeExpense.dateFrom = session.dateFrom;
-      incomeExpense.dateTo = session.dateTo;
-
-      // pick the cashbox account name
-      if (!incomeExpense.accountName && incomeExpense.incomes.length) {
-        incomeExpense.accountName = incomeExpense.incomes[0].label;
-      } else if (!incomeExpense.accountName && incomeExpense.expenses && incomeExpense.expenses.length) {
-        incomeExpense.accountName = incomeExpense.expenses[0].label;
-      } else {
-        incomeExpense.accountName = incomeExpense.accountName;
-      }
-
-      // pick the cashbox account Number
-      if (!incomeExpense.accountNumber && incomeExpense.incomes.length) {
-        incomeExpense.accountNumber = incomeExpense.incomes[0].number;
-      } else if (!incomeExpense.accountNumber && incomeExpense.expenses && incomeExpense.expenses.length) {
-        incomeExpense.accountNumber = incomeExpense.expenses[0].number;
-      } else {
-        incomeExpense.accountNumber = incomeExpense.accountNumber;
-      }
-
-      session.reportType = parseInt(session.reportType, 10);
-
-      if (session.reportType === 1 || session.reportType === 2) {
-        incomeExpense.incomes.forEach((income) => {
-          sumIncome += income.debit;
+function fetchIncome(options) {
```

Comment:
These two functions build the final object (the object which will be used by the handlebar) by adding properties. Separating them seems clear. If I use one function for both of them, I will be supposed to use ```if(type_id === x)``` in many places and the function will be difficult to read.
Feel free to propose me your implementation may be it will work.